platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/EvaluationContext.java

```diff
@@ -0,0 +1,276 @@
+     * @throws CircularEvaluationException if the owner is currently being evaluated in the outer scope
+     */
+    public <R, E extends Exception> R evaluate(EvaluationOwner owner, ScopedEvaluation<? extends R, E> evaluation) throws E {
+        try (ScopeContext ignored = open(owner)) {
+            return evaluation.evaluate();
+        }
+    }
+
+    /**
+     * Runs the {@code evaluation} with the {@code owner} being marked as "evaluating".
+     * If the owner is already being evaluated, returns {@code fallbackValue}.
+     * <p>
+     * Note that fallback value is not used if the evaluation itself throws {@link CircularEvaluationException}, the exception propagates instead.
+     *
+     * @param owner the owner to evaluate
+     * @param fallbackValue the fallback value to return if the owner is already evaluating
+     * @param evaluation the evaluation
+     * @param <R> the type of the result
+     * @param <E> (optional) exception type being thrown by the evaluation
+     * @return the result of the evaluation
+     * @throws E exception from the {@code evaluation} is propagated
+     */
+    public <R, E extends Exception> R tryEvaluate(EvaluationOwner owner, R fallbackValue, ScopedEvaluation<? extends R, E> evaluation) throws E {
+        if (getContext().isInScope(owner)) {
+            return fallbackValue;
+        }
+        // It is possible that the downstream chain itself forms a cycle.
+        // However, it should be its responsibility to be defined in terms of safe evaluation rather than us intercepting the failure here.
+        return evaluate(owner, evaluation);
+    }
+
+    /**
+     * Runs the {@code evaluation} in a nested evaluation context. A nested context allows to re-enter evaluation of the objects that are being evaluated in the enclosed context.
+     * <p>
+     * Use sparingly. In most cases, it is better to rework the call chain to avoid re-evaluating the owner.
+     *
+     * @param evaluation the evaluation
+     * @param <R> the type of the result
+     * @param <E> (optional) exception type being thrown by the evaluation
+     * @return the result of the evaluation
+     * @throws E exception from the {@code evaluation} is propagated
+     */
+    public <R, E extends Exception> R evaluateNested(ScopedEvaluation<? extends R, E> evaluation) throws E {
+        try (ScopeContext ignored = nested()) {
+            return evaluation.evaluate();
+        }
+    }
+
+    private PerThreadContext getContext() {
+        return threadLocalContext.get();
+    }
+
+    private PerThreadContext setContext(PerThreadContext newContext) {
+        threadLocalContext.set(newContext);
+
+        return newContext;
+    }
+
+    private ScopeContext nested() {
+        return setContext(new PerThreadContext(getContext()));
+    }
+
+    private final class PerThreadContext implements ScopeContext {
+        private final Set<EvaluationOwner> objectsInScope = new ReferenceOpenHashSet<>(EXPECTED_MAX_CONTEXT_SIZE);
```

Comment:
🚀 I have a gut feeling that persistent collections would work best for this use-case. I'll run an experiment on a separate branch to check that

Reply:
> 🚀 I have a gut feeling that persistent collections would work best for this use-case. I'll run an experiment on a separate branch to check that

Could be. I used `ReferenceOpenHashSet` because it avoids allocations and GC overhead was noticeable in my earlier experiments.

Reply:
Not really.

I was under the impression that some allocations could be saved but since the context objects are always needed and then reused, their cost is amortized.

FTR, according to the included jmh benchmark:
* replacing `ReferenceArrayList` with `PersistentList` gives about the same performance
* replacing `ReferenceOpenHashSet` with capsule's Immutable set makes it twice as slow